File: 08. Trees/02. Binary Search Tree/01. Search in a BST/main.js
class TreeNode {
    constructor(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// Recursive Search
function searchBST(root, val) {
    if (!root || root.val === val) return root;
    return val < root.val ? searchBST(root.left, val) : searchBST(root.right, val);
}

// Iterative Search
function searchBSTIterative(root, val) {
    while (root && root.val !== val) {
        root = val < root.val ? root.left : root.right;
    }
    return root;
}

// Main Function
function main() {
    let root = new TreeNode(4);
    root.left = new TreeNode(2, new TreeNode(1), new TreeNode(3));
    root.right = new TreeNode(7);

    let target = 2;
    let result = searchBST(root, target);
    console.log(result ? `Found: ${result.val}` : "Not Found");

    result = searchBSTIterative(root, target);
    console.log(result ? `Found (Iterative): ${result.val}` : "Not Found");
}

main();

File: 08. Trees/02. Binary Search Tree/01. Search in a BST/main.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Recursive Search
def search_bst(root, val):
    if not root or root.val == val:
        return root
    return search_bst(root.left, val) if val < root.val else search_bst(root.right, val)

# Iterative Search
def search_bst_iterative(root, val):
    while root and root.val != val:
        root = root.left if val < root.val else root.right
    return root

# Main function
def main():
    root = TreeNode(4)
    root.left = TreeNode(2, TreeNode(1), TreeNode(3))
    root.right = TreeNode(7)

    target = 2
    result = search_bst(root, target)
    print(f"Found: {result.val}" if result else "Not Found")

    result = search_bst_iterative(root, target)
    print(f"Found (Iterative): {result.val}" if result else "Not Found")

if __name__ == "__main__":
    main()
